### PR TITLE
Add telemetry to application

### PR DESCRIPTION
**Pull request by Feiko**
## Description

Added a custom application inside notification every time someone successfully downloaded a firmware package. 

- For this added a Opt Out for sending this telemetry that can be set via environment settings by setting:  "DOTNET_CLI_TELEMETRY_OPTOUT" to "1".
- A custom connectionstring to Applications Insight can be set with the "iConnectionString" field inside "apssetings.json".

## Motivation and Context

Give nanoFramework more information about the adoption of nanoFramework and insight in growth. 

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFirmwareFlasher.Library/FirmwarePackage.cs [hunk 1]
```diff
@@ -342,7 +360,28 @@ internal async Task<ExitCodes> DownloadAndExtractAsync()
                             Console.ForegroundColor = ConsoleColor.White;
                         }
 
+                        //send app insight on successful download
+
+                        string insightConnectionString = _configuration["iConnectionString"];
+                        string optOut = Environment.GetEnvironmentVariable("DOTNET_CLI_TELEMETRY_OPTOUT");
```
Comment (Ellerbach) [inline nanoFirmwareFlasher.Library/FirmwarePackage.cs:366]: Can we rename it NANOFRAMEWORK_TELEMETRY_OPTOUT?

File: nanoFirmwareFlasher.Library/appsettings.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "iConnectionString": "Put iAppKeyConnectionString here!"
```
Comment (Ellerbach) [inline nanoFirmwareFlasher.Library/appsettings.json:2]: Can you add a token and adjust the build pipeline to use a token? We can add the secret into the repo.

File: nanoFirmwareFlasher.Library/FirmwarePackage.cs [hunk 3]
```diff
@@ -342,7 +360,28 @@ internal async Task<ExitCodes> DownloadAndExtractAsync()
...
+
                         stepSuccessful = true;
+
+                        if (!string.IsNullOrEmpty(insightConnectionString) && optOut != "1")
+                        {
+                            TelemetryClient telemetryClient = new TelemetryClient(new Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration()
```
Comment (Ellerbach) [inline nanoFirmwareFlasher.Library/FirmwarePackage.cs:372]: what happens if the key is not correct? Need to use a try/catch?

**Review by Ellerbach — COMMENTED** [review]
Nice, couple of comments

File: nanoFirmwareFlasher.Library/FirmwarePackage.cs (hunk 1 above)
Comment (Feiko) [reply nanoFirmwareFlasher.Library/FirmwarePackage.cs]: Good catch, forgot to change this variable. 
Done.

File: nanoFirmwareFlasher.Library/FirmwarePackage.cs (hunk 3 above)
Comment (Feiko) [reply nanoFirmwareFlasher.Library/FirmwarePackage.cs:372]: It is already in a try catch, loop see line 386. The code was tested using a faulty connectionstring, when this fails the rest of the program continues as it should.

File: nanoFirmwareFlasher.Library/appsettings.json (hunk 2 above)
Comment (Feiko) [reply nanoFirmwareFlasher.Library/appsettings.json:2]: @josesimoes could you add the secret token/connectionstring? I don't have access to this connectionstring.

File: azure-pipelines.yml [hunk 4]
```diff
@@ -327,6 +327,25 @@ jobs:
...
+# Replace the intrument key
+    - powershell: |
+            . ./common.ps1
+            $fileToReplace = "$(System.DefaultWorkingDirectory)/nanoFirmwareFlasher.Library/appsettings.json"
+            $sourceString = "INTRUMENT_KEY"
+            $instrumentKey = "$(InstrumentKey)"
+
+            Find-ReplaceInFile -filePath $fileToReplace -sourceString $sourceString -targetString $instrumentKey
+        displayName: Replace intrument key
+        condition: >-
```
Comment (Ellerbach) [inline azure-pipelines.yml:339]: conditions needs to be adjusted (I'll take care of this, it's a note for myself)

**Review by josesimoes — APPROVED** [review]
LGTM!
I suppose that this can be merged now... @Ellerbach ?

**Pull request merged**